File: validParentheses/validParentheses.cpp
#include <iostream>
#include <stack>
#include <unordered_map>
using namespace std;

class Solution {
public:
    bool isValid(string s) {
        stack<char> stk;
        unordered_map<char, char> m;
        m.insert(make_pair(')', '('));
        m.insert(make_pair('}', '{'));
        m.insert(make_pair(']', '['));
        for (char c : s) {
            if (c == ')' || c == '}' || c == ']') {
                if (stk.empty() || stk.top() != m[c]) {
                    return false;
                } else {
                    stk.pop();
                }
            } else {
                stk.push(c);
            }
        }
        return stk.empty();
    }
};

int main()
{
  Solution s;
  cout << s.isValid(string("{}{}{}{}[][]")) << endl;
  return 0;
}

File: validParentheses/validParentheses.java
import java.util.*;

class ValidParentheses {
    public boolean isValid(String s) {
        Map<Character, Character> m = new HashMap<>();
        m.put('(', ')');
        m.put('[', ']');
        m.put('{', '}');
        Stack<Character> stk = new Stack<>();
        for (char c : s.toCharArray()) {
            if (c == ')' || c == ']' || c == '}') {
                if (!stk.isEmpty() && m.containsKey(stk.peek()) && m.get(stk.peek()) == c) {
                    stk.pop();
                } else {
                    return false;
                }
            } else {
                stk.push(c);
            }
        }
        return stk.isEmpty();
    }
	public static void main(String[] args) {
		ValidParentheses s = new ValidParentheses();
		String str = "{[]}";
		System.out.println(s.isValid(str));
	}
}
